coil-gif/src/nonAndroidMain/kotlin/coil3/gif/AnimatedSkiaImageDecoder.kt

```diff
@@ -0,0 +1,202 @@
+package coil3.gif
+
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableIntStateOf
+import androidx.compose.runtime.setValue
+import coil3.Canvas
+import coil3.Image
+import coil3.ImageLoader
+import coil3.decode.DecodeResult
+import coil3.decode.DecodeUtils
+import coil3.decode.Decoder
+import coil3.decode.ImageSource
+import coil3.fetch.SourceFetchResult
+import coil3.request.Options
+import kotlin.time.TimeSource
+import okio.BufferedSource
+import okio.use
+import org.jetbrains.skia.AnimationFrameInfo
+import org.jetbrains.skia.Bitmap
+import org.jetbrains.skia.Codec
+import org.jetbrains.skia.ColorAlphaType
+import org.jetbrains.skia.ColorInfo
+import org.jetbrains.skia.ColorSpace
+import org.jetbrains.skia.ColorType
+import org.jetbrains.skia.Data
+import org.jetbrains.skia.Image as SkiaImage
+import org.jetbrains.skia.ImageInfo
+
+/**
+ * A [Decoder] that uses Skia to decode animated images (GIF, WebP).
+ *
+ * @param prerenderFrames If true, prerender all frames when the image is decoded.
+ * This means the image will consume more memory, and will only start playing after it has been
+ * fully buffered, but will also play more smoothly.
+ */
+class AnimatedSkiaImageDecoder(
+    private val source: ImageSource,
+    private val prerenderFrames: Boolean = true,
```

Comment:
I think we'll need to tweak this to something like `bufferedFramesCount` (pass 0 for disabled, Int.MAX_VALUE to buffer everything) since we'll need to update the implementation to buffer a few frames up front then buffer the rest just in time before we need to draw them.

We'll want to buffer X frames during the decode process then figure out how to get a `CoroutineScope` so we can decode the rest right after `AnimatedSkiaImage.draw` is called.